File: src/main/java/za/co/dubedivine/networks/controller/TagController.java
package za.co.dubedivine.networks.controller;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import za.co.dubedivine.networks.model.Tag;
import za.co.dubedivine.networks.model.User;
import za.co.dubedivine.networks.model.repository.TagRepository;
import za.co.dubedivine.networks.model.repository.UserRepository;
import za.co.dubedivine.networks.model.responseEntity.StatusResponseEntity;

import java.util.List;

/**
 * Created by divine on 2017/08/11.
 */
@RestController
@RequestMapping("/tags")
public class TagController {

    private final UserRepository userRepository;
    private final TagRepository tagRepository;

    public TagController(TagRepository tagRepository,
                         UserRepository userRepository) {
        this.tagRepository = tagRepository;
        this.userRepository = userRepository;
    }



    @GetMapping("/all")
    public List<Tag> getAll() {
        List<Tag> tags = this.tagRepository.findAll();
        return tags;
    }

    @PostMapping("/{t_id}/subscribe/{u_id}")
    public ResponseEntity<StatusResponseEntity> subscribeToTag(@PathVariable("t_id") String tagId,
                                                               @PathVariable("u_id") String userId) {

        //function for a user  to listen to a particular TAG
        Tag tag = tagRepository.findOne(tagId);
        User user = userRepository.findOne(userId);
        if (user != null && tag != null) {

            List<Tag> tags = user.getTags();
            tags.add(tag);
            user.setTags(tags);

            userRepository.save(user);
            return
                    new ResponseEntity<>(
                            new StatusResponseEntity(true,
                                    "you now listening on tag"), HttpStatus.CREATED);
        } else {

            return
                    new ResponseEntity<>(
                            new StatusResponseEntity(false,
                                    "you now listening on tag"), HttpStatus.CREATED);
        }
    }


    @PutMapping
    public void insert(@RequestBody Tag tag) {
        this.tagRepository.insert(tag);
    }

    @PostMapping
    public void update(@RequestBody Tag tag) {
        tagRepository.save(tag);
    }

    @DeleteMapping("/{id}")
    public void delete(@PathVariable("id") String id) {
        //todo: should also go the tags part and then delete the reference
        tagRepository.delete(id);
    }
}

File: src/main/java/za/co/dubedivine/networks/controller/TagController.kt
package za.co.dubedivine.networks.controller

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import za.co.dubedivine.networks.model.Tag
import za.co.dubedivine.networks.model.elastic.ElasticTag
import za.co.dubedivine.networks.repository.TagRepository
import za.co.dubedivine.networks.repository.UserRepository
import za.co.dubedivine.networks.model.responseEntity.StatusResponseEntity
import za.co.dubedivine.networks.services.elastic.ElasticTagService
import za.co.dubedivine.networks.util.KUtils

/**
 * Created by divine on 2017/08/11.
 */
@RestController
@RequestMapping("/tags")
class TagController(private val tagRepository: TagRepository,
                    private val userRepository: UserRepository,
                    private val elasticTagService: ElasticTagService) {


    val getAll: List<Tag>
        @GetMapping
        get() = tagRepository.findAll()

    @PostMapping("/{t_id}/subscribe/{u_id}")
    fun subscribeToTag(@PathVariable("t_id") tagId: String,
                       @PathVariable("u_id") userId: String): ResponseEntity<StatusResponseEntity> {

        //function for a user  to listen to a particular TAG
        val tag = tagRepository.findOne(tagId)
        val user = userRepository.findOne(userId)
        if (user != null && tag != null) {

            val tags = user.tags
            tags.add(tag)
            user.tags = tags

            userRepository.save(user)
            return ResponseEntity(
                    StatusResponseEntity(true,
                            "you now listening on tag"), HttpStatus.CREATED)
        } else {

            return ResponseEntity(
                    StatusResponseEntity(false,
                            "you now listening on tag"), HttpStatus.CREATED)
        }
    }


    @PutMapping
    fun insert(@RequestBody tag: Tag) {
        this.tagRepository.insert(tag)
    }

    @PostMapping
    fun update(@RequestBody tag: Tag) {
        tagRepository.save(tag)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable("id") id: String) {
        //todo: should also go the tags part and then delete the reference
        tagRepository.delete(id)
    }

    @GetMapping("/suggest")
    fun suggestTags(@RequestParam("tag") tag: String): List<ElasticTag> {
        println("here is the passed in tag bro $tag")
        return elasticTagService.suggestTag(KUtils.cleanTag(tag))
    }
}
